File: php-backend/src/Services/CatalogResolver.php
<?php

declare(strict_types=1);

namespace App\Services;

use App\Database;

/**
 * 逻辑 file/sheet/列头 → 物理表字段，与 Python 版 services/catalog_resolver.py 对等。
 */
final class CatalogResolver
{
    public static function hasCatalog(int $dataSourceId): bool
    {
        return (bool) Database::fetchValue(
            'SELECT id FROM catalog_files WHERE data_source_id = ? AND is_active = 1 LIMIT 1',
            [$dataSourceId]
        );
    }

    /** @return array<int, array{file_name: string, keyword: string}> */
    public static function listCatalogFiles(int $dataSourceId): array
    {
        $rows = Database::fetchAll(
            'SELECT file_name, keyword FROM catalog_files WHERE data_source_id = ? AND is_active = 1 ORDER BY id',
            [$dataSourceId]
        );
        return array_map(fn($r) => ['file_name' => $r['file_name'], 'keyword' => $r['keyword']], $rows);
    }

    /** @return string[] */
    public static function listCatalogSheets(int $dataSourceId, string $fileKeyword): array
    {
        $rows = Database::fetchAll(
            'SELECT DISTINCT cs.sheet_name FROM catalog_sheets cs
             JOIN catalog_files cf ON cs.file_id = cf.id
             WHERE cf.data_source_id = ? AND cf.keyword = ? AND cf.is_active = 1 AND cs.is_active = 1',
            [$dataSourceId, $fileKeyword]
        );
        $names = array_map(fn($r) => (string) $r['sheet_name'], $rows);
        sort($names);
        return $names;
    }

    /** @return string[] */
    public static function listCatalogColumns(int $dataSourceId, string $fileKeyword, string $sheetName): array
    {
        $rows = Database::fetchAll(
            'SELECT cc.header_name FROM catalog_columns cc
             JOIN catalog_sheets cs ON cc.sheet_id = cs.id
             JOIN catalog_files cf ON cs.file_id = cf.id
             WHERE cf.data_source_id = ? AND cf.keyword = ? AND cs.sheet_name = ?
               AND cf.is_active = 1 AND cs.is_active = 1 AND cc.is_active = 1',
            [$dataSourceId, $fileKeyword, $sheetName]
        );
        $names = [];
        foreach ($rows as $r) {
            if ($r['header_name'] !== null && $r['header_name'] !== '') {
                $names[$r['header_name']] = true;
            }
        }
        $out = array_keys($names);
        sort($out);
        return $out;
    }
}

File: app/services/catalog_resolver.py
"""逻辑 file/sheet/列头 → 物理表字段。"""

from __future__ import annotations

import re
from dataclasses import dataclass

from sqlalchemy.orm import Session

from app.models import CatalogColumn, CatalogFile, CatalogSheet


def catalog_file_display_label(file: CatalogFile) -> str:
    """UI 下拉用短名：优先 file_label / keyword，不从完整 file_name 截断。"""
    lbl = (file.file_label or "").strip()
    if lbl and lbl != file.file_name and not lbl.lower().endswith((".xlsx", ".xls")):
        return lbl
    kw = (file.keyword or "").strip()
    if kw:
        return kw
    name = re.sub(r"\.xlsx?$", "", file.file_name or "", flags=re.I)
    name = re.sub(r"_\d{8}(?:-\d{8})?$", "", name)
    parts = [p for p in name.split("_") if p]
    return parts[-1] if parts else (name or "文件")


@dataclass
class ResolvedColumn:
    file_id: int
    file_keyword: str
    file_name: str
    sheet_name: str
    fact_table: str
    header_name: str
    db_column: str
    column_aliases: list[str]


def has_catalog(db: Session, data_source_id: int) -> bool:
    return (
        db.query(CatalogFile.id)
        .filter(CatalogFile.data_source_id == data_source_id, CatalogFile.is_active.is_(True))
        .first()
        is not None
    )


def resolve_column(
    db: Session,
    data_source_id: int,
    file_keyword: str | None,
    sheet_name: str,
    header_name: str,
) -> ResolvedColumn | None:
    q = (
        db.query(CatalogColumn, CatalogSheet, CatalogFile)
        .join(CatalogSheet, CatalogColumn.sheet_id == CatalogSheet.id)
        .join(CatalogFile, CatalogSheet.file_id == CatalogFile.id)
        .filter(
            CatalogFile.data_source_id == data_source_id,
            CatalogFile.is_active.is_(True),
            CatalogSheet.is_active.is_(True),
            CatalogColumn.is_active.is_(True),
            CatalogSheet.sheet_name == sheet_name,
        )
    )
    if file_keyword:
        q = q.filter(CatalogFile.keyword == file_keyword)

    for col, sheet, file in q.all():
        names = [col.header_name, *(col.column_aliases or [])]
        if header_name in names:
            return ResolvedColumn(
                file_id=file.id,
                file_keyword=file.keyword,
                file_name=file.file_name,
                sheet_name=sheet.sheet_name,
                fact_table=sheet.fact_table,
                header_name=col.header_name,
                db_column=col.db_column,
                column_aliases=col.column_aliases or [],
            )
    return None


def list_catalog_files(db: Session, data_source_id: int) -> list[dict]:
    rows = (
        db.query(CatalogFile)
        .filter(CatalogFile.data_source_id == data_source_id, CatalogFile.is_active.is_(True))
        .order_by(CatalogFile.id)
        .all()
    )
    return [
        {
            "file_name": r.file_name,
            "keyword": r.keyword,
            "file_label": r.file_label,
            "label": catalog_file_display_label(r),
        }
        for r in rows
    ]


def list_catalog_sheets(db: Session, data_source_id: int, file_keyword: str) -> list[str]:
    rows = (
        db.query(CatalogSheet.sheet_name)
        .join(CatalogFile, CatalogSheet.file_id == CatalogFile.id)
        .filter(
            CatalogFile.data_source_id == data_source_id,
            CatalogFile.keyword == file_keyword,
            CatalogFile.is_active.is_(True),
            CatalogSheet.is_active.is_(True),
        )
        .all()
    )
    return sorted({name for (name,) in rows})


def list_catalog_columns(
    db: Session, data_source_id: int, file_keyword: str, sheet_name: str
) -> list[str]:
    rows = (
        db.query(CatalogColumn.header_name)
        .join(CatalogSheet, CatalogColumn.sheet_id == CatalogSheet.id)
        .join(CatalogFile, CatalogSheet.file_id == CatalogFile.id)
        .filter(
            CatalogFile.data_source_id == data_source_id,
            CatalogFile.keyword == file_keyword,
            CatalogSheet.sheet_name == sheet_name,
            CatalogFile.is_active.is_(True),
            CatalogSheet.is_active.is_(True),
            CatalogColumn.is_active.is_(True),
        )
        .all()
    )
    return sorted({name for (name,) in rows if name})
